homeassistant/helpers/aiohttp_client.py

```diff
@@ -70,6 +70,19 @@ def async_create_clientsession(
         **kwargs,
     )
 
+    async def patched_close() -> None:
+        """Mock close to avoid integrations closing our session."""
+
+    orig_session_close = clientsession.close
+    clientsession.close = patched_close  # type: ignore
+
+    @callback
+    def _async_restore_close(event: Event) -> None:
+        """Restore session close method."""
+        clientsession.close = orig_session_close  # type: ignore
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_CLOSE, _async_restore_close)
```

Comment:
`_async_register_clientsession_shutdown` also executes on close and now you're in a race condition if you can restore in time before actual closing. You should pass the original close function to `_async_register_clientsession_shutdown` and don't restore it.

Reply:
Thanks!